lib/datadog/di/code_tracker.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+module Datadog
+  module DI
+    # Tracks loaded Ruby code by source file and maintains a map from
+    # source file to the loaded code (instruction sequences).
+    # Also arranges for code in the loaded files to be instrumented by
+    # line probes that have already been received by the library.
+    #
+    # The loaded code is used to target line trace points when installing
+    # line probes which dramatically improves efficiency of line trace points.
+    #
+    # Note that, since most files will only be loaded one time (via the
+    # "require" mechanism), the code tracker needs to be global and not be
+    # recreated when the DI component is created.
+    #
+    # @api private
+    class CodeTracker
+      def initialize
+        @registry = Hash.new
+        @trace_point_lock = Mutex.new
+        @registry_lock = Hash.new
+      end
+
+      def start
```

Comment:
If this method is not called at a high frequency (which it doesn't seem like it is), I believe to make it truly tread safe we have to wrap the whole execution in a mutex.

Reply:
I already have a lock taken for all reads and writes of instance variables and  I believe  I account for e.g. state changing during constructor execution. Can you elaborate on what situation you think is not properly covered by existing locking?